### PR TITLE
DO NOT MERGE: [stdlib] CharacterView not own SliceType, and drop Unicode 8 tries (finally)

### PR DESCRIPTION
**Pull request by milseman**
[stdlib] String.CharacterView.SubSequence is Substring.CharacterView
    
This changes String.CharacterView from being its own slice type to
instead use Substring.CharacterView. In the process, this updates
Substring.CharacterView to be RangeReplaceable.

This aligns with the goal of introducing Substring as a separate type
to avoid accidental memory "leaks".

---

[stdlib] Drop Unicode 8 tries in stdlib (entirely)
    
This drops the last vestage of Unicode 8 tries from the standard
library. Switches everything over to use ICU.

### CODE REVIEW

**Comment by milseman** [comment]
@swift-ci please test

**Comment by milseman** [comment]
This failed due to infrastructure problems, but didn't report back.

**Comment by milseman** [comment]
@swift-ci please test

**Comment by shahmishal** [comment]
@milseman If you see it again can you please send me an email with Job URL. Thanks!

**Comment by milseman** [comment]
Will do!

**Comment by milseman** [comment]
So yeah... this is totally a source break for Swift 4 code. I'll need to do some `@available` hacky judo.

File: stdlib/public/core/Substring.swift.gyb [hunk 1]
```diff
@@ -384,7 +384,7 @@ extension StringProtocol {
 %   ('characters', 'Character')
 % ]:
 %   View = ViewPrefix + 'View'
-%   RangeReplaceable = 'RangeReplaceable' if property == 'unicodeScalars' else ''
+%   RangeReplaceable = 'RangeReplaceable' if property == 'unicodeScalars' or property == 'characters' else ''
```
Comment (moiseev) [inline stdlib/public/core/Substring.swift.gyb]: Minor nitpick: `if property in ["unicodeScalars", "characters"]`.

**Review by moiseev — COMMENTED** [review]
LGTM

File: stdlib/public/core/StringCharacterView.swift [hunk 2]
```diff
@@ -739,6 +741,12 @@ extension String.CharacterView {
   ///
   /// - Complexity: O(*n*) if the underlying string is bridged from
   ///   Objective-C, where *n* is the length of the string; otherwise, O(1).
+  @available(swift, introduced: 4.0)
+  public subscript(bounds: Range<Index>) -> Substring.CharacterView {
```
Comment (natecook1000) [inline stdlib/public/core/StringCharacterView.swift]: What's the story around substrings and bridging? Do we automatically bridge an objc string when it's substring'd, or can a substring be `NSString`-backed?

Comment (moiseev) [inline stdlib/public/core/StringCharacterView.swift]: Right now when the bridging happens, a new `String` gets created from a substring [here](https://github.com/apple/swift/blob/41b00c57ab90c176e9d24098f839b315a2812a50/stdlib/public/SDK/Foundation/String.swift#L63).

File: test/stdlib/StringCompatibility.swift [hunk 3]
```diff
@@ -356,5 +356,20 @@ Tests.test("LosslessStringConvertible/force unwrap/\(swift)") {
 }
 #endif
 
+#if !swift(>=4)
+Tests.test("popFirst") {
```
Comment (moiseev) [inline test/stdlib/StringCompatibility.swift]: Just a little comment, when the test is run in both Swift 3 and Swift 4 modes, it is quite useful to add a swift version in a test name. Like the following.

```swift
Tests.test("LosslessStringConvertible/force unwrap/\(swift)") {
```

It is not particularly important for this test right no as it only exists in Swift 3 world, but refactoring happens...

**Comment by milseman** [comment]
`@available` judo is insufficient, as it cannot version associated types. This leaves us in the situation where either: we leak, we break source, or we deprecate CharacterView asap (post 4.0)

I'll split this PR into multiple ones to tackle the other issues unrelated to String.CharacterView, and mock up what the third option is like.

**Pull request closed without merging**